core/src/test/scala/org/apache/spark/rdd/RDDSuite.scala

```diff
@@ -1176,6 +1176,24 @@ class RDDSuite extends SparkFunSuite with SharedSparkContext {
     }.collect()
   }
 
+  test("SPARK-27666: Do not release lock while TaskContext already completed") {
+    val rdd = sc.parallelize(Range(0, 10), 1).cache()
+    // validate cache
+    rdd.collect()
+    rdd.mapPartitions { iter =>
+      val t = new Thread(() => {
+        while (iter.hasNext) {
+          iter.next()
+          Thread.sleep(100)
+        }
+      })
+      t.setDaemon(false)
+      t.start()
+      Iterator(0)
+    }.collect()
+    Thread.sleep(10 * 150)
```

Comment:
Can we use a more reliable way to test it? We can set up a `CountDownLatch`, and count it down at the end of the thread. Then we wait for the count down at the end of the test.

Reply:
`CountDownLatch` can't be serialized in Task, so it doesn't work.

Reply:
Why not add a TaskCompletionListener and wait for the listener to be triggered? Something like:
```
      eventually(timeout(10.seconds)) {
          assert(// Task completion triggered.)
      }
```

Reply:
IIUC, `TaskCompletionListener` will be called after `collect()` done, but thread `t` will be still running at that time.

Reply:
Make sense. Also curious how did the magic number `10 * 150` come out?

Reply:
Thread.sleep(10 * 150) > iter.size * Thread(100)

Reply:
we shouldn't use sleep in tests, as the test will become flaky sooner or later. If `CountDownLatch` doesn't work, can we use Spark `Accumulator` as signals?

Reply:
I tried `Accumulator` previously, it looks like child thread continuously add acc by 1 every 100ms. And test thread waits until acc reaches 10. But it doesn't work, either. Because we want the signal comes from that child thread, but acc always comes with finished task. Unfortunately, in this case, task finished before that child thread.

Reply:
If this kind of end-to-end test doesn't work without `sleep`, how about we have an unit test for `releaseLock` behavior when task context is completed?

Reply:
How about this way eba8e96 ? I test it with/without this PR, and it works.